keras_rs/src/layers/modeling/dot_interaction.py

```diff
@@ -0,0 +1,131 @@
+from typing import Any, List, Tuple
+
+import keras
+from keras import ops
+
+from keras_rs.src import types
+from keras_rs.src.api_export import keras_rs_export
+
+
+@keras_rs_export("keras_rs.layers.DotInteraction")
+class DotInteraction(keras.layers.Layer):
+    """Dot interaction layer present in DLRM.
+
+    This layer computes distinct dot product ("feature interactions") for every
+    pair of features. If `self_interaction` is True, we calculate dot products
+    of the form `dot(e_i, e_j)` for `i <= j`, and `dot(e_i, e_j)` for `i < j`,
+    otherwise. `e_i` represents features. The layer can be used to build
+    the DLRM model.
+
+    Args:
+        self_interaction: bool. Indicates whether features should
+            "self-interact". If it is True, then the diagonal entries of the
+            interaction matrix are also taken.
+        skip_gather: bool. If it's set, then the upper triangular part of the
+            interaction matrix is set to 0. The output will be of shape
+            `[batch_size, num_features * num_features]` of which half of the
+            entries will be zeros. Otherwise, the output will be only the lower
+            triangular part of the interaction matrix. The latter saves space
+            but is much slower.
+
+    References:
+    - [M. Naumov et al.](https://arxiv.org/abs/1906.00091)
+    """
+
+    def __init__(
+        self,
+        self_interaction: bool = False,
+        skip_gather: bool = False,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(**kwargs)
+
+        # Passed args.
+        self.self_interaction = self_interaction
+        self.skip_gather = skip_gather
+
+    def call(self, inputs: List[types.Tensor]) -> types.Tensor:
+        """Forward pass of the dot interaction layer.
+
+        Args:
+            inputs: list. Every element in the list represents a feature tensor
+                of shape `[batch_size, feature_dim]`. All tensors in the list
+                must have the same shape.
+
+        Returns:
+            Tensor representing feature interactions. The shape of the tensor is
+            `[batch_size, k]` where `k` is
+            `num_features * num_features` if `skip_gather` is `True`. Otherwise,
+            `k` is `num_features * (num_features + 1) / 2` if
+            `self_interaction` is `True`, and
+            `num_features * (num_features - 1) / 2` if not.
+        """
+
+        # A `ValueError` will be raised if the feature tensors inside the list
+        # are not of equal shape. So, we do not need to handle the error here.
+        features = ops.array(inputs)
```

Comment:
This or `ops.concatenate([ops.expand_dims(x, axis=1) for x in inputs], axis=1)`?